src/history/hs.c

```diff
@@ -838,19 +824,50 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
             if (prev_upd != NULL && prev_upd->start_ts < upd->start_ts)
                 enable_reverse_modify = false;
 
-            /* Find the first update without timestamp. */
-            if (first_non_ts_upd == NULL && upd->start_ts == WT_TS_NONE) {
-                first_non_ts_upd = upd;
-            } else if (first_non_ts_upd != NULL && upd->start_ts != WT_TS_NONE) {
+            if (first_non_ts_upd == NULL) {
+                /* Find the first update without timestamp. */
+                if (upd->start_ts == WT_TS_NONE)
+                    first_non_ts_upd = upd;
+                else if (min_insert_ts < upd->start_ts) {
+                    /*
+                     * Only fix out of order timestamp if we haven't seen an update without
+                     * timestamp.
+                     *
+                     * Resolved prepared updates will lose their durable timestamp here. This is a
+                     * wrinkle in our handling of out-of-order updates.
+                     */
+                    if (upd->start_ts != upd->durable_ts) {
+                        WT_ASSERT(session, min_insert_ts < upd->durable_ts);
+                        WT_STAT_CONN_INCR(session, cache_hs_order_lose_durable_timestamp);
+                    }
+                    __wt_verbose(session, WT_VERB_TIMESTAMP,
+                      "fixing out-of-order updates during insertion; start_ts=%s, "
+                      "durable_start_ts=%s, "
+                      "min_insert_ts=%s",
+                      __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
+                      __wt_timestamp_to_string(upd->durable_ts, ts_string[1]),
+                      __wt_timestamp_to_string(min_insert_ts, ts_string[2]));
+                    upd->start_ts = upd->durable_ts = min_insert_ts;
+                    WT_STAT_CONN_INCR(session, cache_hs_order_fixup_insert);
+                } else
+                    min_insert_ts = upd->start_ts;
+            } else if (upd->start_ts != WT_TS_NONE) {
                 /*
                  * Don't insert updates with timestamps after updates without timestamps to the
                  * history store.
                  */
-                F_SET(upd, WT_UPDATE_MASKED_BY_NON_TS_UPDATE);
+                F_SET(upd, WT_UPDATE_OBSOLETE);
                 if (F_ISSET(upd, WT_UPDATE_HS))
                     ts_updates_in_hs = true;
             }
 
+            /*
+             * No need to continue if we see the first self contained value after the first globally
+             * visible value.
+             */
+            if (first_globally_visible_upd != NULL && WT_UPDATE_DATA_VALUE(upd))
+                break;
```

Comment:
This short circuit means we can't rely on updates older than this being marked with `WT_UPDATE_OBSOLETE` - is that OK?

Should we be careful at all about updates that come into this function already marked as obsolete?

Reply:
That is OK as they are not pushed to the stack. We only care about updates pushed to the stack after this loop.